File: app/src/main/java/com/medziku/plantwatering/bluetooth/BluetoothDevicePickFragment.java
package com.medziku.plantwatering.bluetooth;

import android.app.Dialog;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AlertDialog;
import android.util.Log;
import android.widget.ListView;
import android.widget.Toast;

import com.medziku.plantwatering.R;

import java.util.Set;

public class BluetoothDevicePickFragment extends DialogFragment {
    private static final String TAG = BluetoothDevicePickFragment.class.getName();

    private BluetoothAdapter mBluetoothAdapter = BluetoothAdapter.getDefaultAdapter();

    private final BroadcastReceiver mReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            final String action = intent.getAction();

            if (action.equals(BluetoothAdapter.ACTION_STATE_CHANGED)) {
                final int state = intent.getIntExtra(BluetoothAdapter.EXTRA_STATE, BluetoothAdapter.ERROR);
//                updateUiOnBTStateChanged(state);
            } else if (BluetoothAdapter.ACTION_DISCOVERY_STARTED.equals(action)) {
                Log.d(TAG, "Discovery Started");
            } else if (BluetoothAdapter.ACTION_DISCOVERY_FINISHED.equals(action)) {
                Log.d(TAG, "Discovery Finished");
            } else if (BluetoothDevice.ACTION_FOUND.equals(action)) {
                // Bluetooth device found
                BluetoothDevice device = intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE);
                Toast.makeText(getActivity().getApplicationContext(),
                        "Found device " + device.getName(), Toast.LENGTH_SHORT)
                        .show();
            }
        }
    };

    public static BluetoothDevicePickFragment newInstance() {
        BluetoothDevicePickFragment frag = new BluetoothDevicePickFragment();
        return frag;
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        final BluetoothDeviceAdapter mBluetoothArrayAdapter = new BluetoothDeviceAdapter(this.getActivity(),
                mBluetoothAdapter.getName(),
                mBluetoothAdapter.getAddress());

        ListView modeList = new ListView(this.getActivity());
        modeList.setAdapter(mBluetoothArrayAdapter);

        return new AlertDialog.Builder(getActivity())
                .setIcon(R.drawable.caution)
                .setView(modeList)
                .create();
    }

    @Override
    public void onStart() {
        super.onStart();
        IntentFilter filter = new IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED);
        filter.addAction(BluetoothDevice.ACTION_FOUND);
        filter.addAction(BluetoothAdapter.ACTION_DISCOVERY_STARTED);
        filter.addAction(BluetoothAdapter.ACTION_DISCOVERY_FINISHED);
        getActivity().registerReceiver(mReceiver, filter);

        Set<BluetoothDevice> pairedDevices = mBluetoothAdapter.getBondedDevices();
        BluetoothDeviceAdapter mBluetoothArrayAdapter = new BluetoothDeviceAdapter(getActivity(),
                mBluetoothAdapter.getName(),
                mBluetoothAdapter.getAddress());
        // If there are paired devices
        if (pairedDevices.size() > 0) {//TODO move to dialog display function
            // Loop through paired devices
            for (BluetoothDevice device : pairedDevices) {
                // Add the name and address to an array adapter to show in a ListView
                mBluetoothArrayAdapter.addBoundedDevice(device);
            }
        }
    }


}

File: android-app/src/main/kotlin/com/medziku/plantwatering/bluetooth/BluetoothDevicePickFragment.kt
package com.medziku.plantwatering.bluetooth

import android.app.Dialog
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v7.app.AlertDialog
import android.util.Log
import android.widget.ListView
import android.widget.Toast
import com.medziku.plantwatering.R

class BluetoothDevicePickFragment : DialogFragment() {

    companion object {
        private val TAG: String = BluetoothDevicePickFragment::class.java.name

        fun newInstance(): BluetoothDevicePickFragment {
            return BluetoothDevicePickFragment()
        }
    }

    private val bluetoothAdapter: BluetoothAdapter = BluetoothAdapter.getDefaultAdapter()

    private val broadcastReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            when (intent.action) {
                BluetoothAdapter.ACTION_STATE_CHANGED -> {
                    val state = intent.getIntExtra(BluetoothAdapter.EXTRA_STATE, BluetoothAdapter.ERROR)
                    Log.d(TAG, "BT state changed to $state")
                    //updateUiOnBTStateChanged(state)
                }
                BluetoothAdapter.ACTION_DISCOVERY_STARTED -> {
                    Log.d(TAG, "Discovery Started")
                }
                BluetoothAdapter.ACTION_DISCOVERY_FINISHED -> {
                    Log.d(TAG, "Discovery Finished")
                }
                BluetoothDevice.ACTION_FOUND -> { // Bluetooth device found
                    val device: BluetoothDevice = intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE)
                    Toast.makeText(activity.applicationContext,
                            "Found device " + device.name, Toast.LENGTH_SHORT)
                            .show()
                }
            }
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val bluetoothDeviceAdapter = BluetoothDeviceAdapter(activity, deviceName = bluetoothAdapter.name, deviceAddress = bluetoothAdapter.address)
        val listViewCompat = ListView(activity)
        listViewCompat.adapter = bluetoothDeviceAdapter
        return AlertDialog.Builder(activity)
                .setIcon(R.drawable.caution)
                .setView(listViewCompat)
                .create()
    }

    override fun onStart() {
        super.onStart()
        val filter = IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED)
        filter.addAction(BluetoothDevice.ACTION_FOUND)
        filter.addAction(BluetoothAdapter.ACTION_DISCOVERY_STARTED)
        filter.addAction(BluetoothAdapter.ACTION_DISCOVERY_FINISHED)
        activity.registerReceiver(broadcastReceiver, filter)

        val pairedDevices: Set<BluetoothDevice> = bluetoothAdapter.bondedDevices
        val mBluetoothArrayAdapter = BluetoothDeviceAdapter(activity,
                deviceName = bluetoothAdapter.name,
                deviceAddress = bluetoothAdapter.address)
        // If there are paired devices
        if (pairedDevices.isNotEmpty()) { //TODO move to dialog display function
            // Loop through paired devices
            for (device in pairedDevices) { // Add the name and address to an array adapter to show in a ListView
                mBluetoothArrayAdapter.addBoundedDevice(device)
            }
        }
    }

}